src/bin/pg_autoctl/pghba.c

```diff
@@ -155,15 +202,176 @@ pghba_ensure_host_rule_exists(const char *hbaFilePath,
+ * group, presumably.
+ *
+ * Each node in the array needs two rules:
+ *
+ *  host(ssl) replication "pgautofailover_replicator" nodename/ip trust
+ *  host(ssl) "dbname"    "pgautofailover_replicator" nodename/ip trust
+ */
+bool
+pghba_ensure_host_rules_exist(const char *hbaFilePath,
+							  NodeAddressArray *nodesArray,
+							  bool ssl,
+							  const char *database,
+							  const char *username,
+							  const char *authenticationScheme)
+{
+	PQExpBuffer newHbaContents = createPQExpBuffer();
+	char *currentHbaContents = NULL;
+	long currentHbaSize = 0L;
+	char *includeLine = NULL;
+
+	int nodeIndex = 0;
+
+	if (newHbaContents == NULL)
+	{
+		log_error("Failed to allocate memory");
+		return false;
+	}
+
+	if (!read_file(hbaFilePath, &currentHbaContents, &currentHbaSize))
+	{
+		/* read_file logs an error */
+		return false;
+	}
+
+	/* always begin with the existing HBA file */
+	appendPQExpBufferStr(newHbaContents, currentHbaContents);
+
+	for (nodeIndex = 0; nodeIndex < nodesArray->count; nodeIndex++)
+	{
+		NodeAddress *node = &(nodesArray->nodes[nodeIndex]);
+		PQExpBuffer hbaLineReplicationBuffer = createPQExpBuffer();
+		PQExpBuffer hbaLineDatabaseBuffer = createPQExpBuffer();
+
+		int hbaLinesIndex = 0;
+		PQExpBuffer hbaLines[] = {
+			hbaLineReplicationBuffer,
+			hbaLineDatabaseBuffer,
+			NULL
+		};
+
+		if (hbaLineReplicationBuffer == NULL ||
+			hbaLineDatabaseBuffer == NULL)
+		{
+			log_error("Failed to allocate memory");
+			return false;
+		}
+
+		log_debug("pghba_ensure_host_rules_exist: %d %s:%d",
+				  node->nodeId, node->host, node->port);
+
+		if (!pghba_append_rule_to_buffer(hbaLineReplicationBuffer,
+										 ssl,
+										 HBA_DATABASE_REPLICATION,
+										 NULL,
+										 username,
+										 node->host,
+										 authenticationScheme))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		if (!pghba_append_rule_to_buffer(hbaLineDatabaseBuffer,
+										 ssl,
+										 HBA_DATABASE_DBNAME,
+										 database,
+										 username,
+										 node->host,
+										 authenticationScheme))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		log_info("Ensuring HBA rules for node %d (%s:%d)",
+				 node->nodeId, node->host, node->port);
+
+		for (hbaLinesIndex = 0; hbaLines[hbaLinesIndex] != NULL; hbaLinesIndex++)
+		{
+			PQExpBuffer hbaLineBuffer = hbaLines[hbaLinesIndex];
+
+			log_debug("Ensuring the HBA file \"%s\" contains the line: %s",
+					  hbaFilePath, hbaLineBuffer->data);
+
+			includeLine = strstr(currentHbaContents, hbaLineBuffer->data);
+
+			/*
+			 * If the rule was found and it starts on a new line. We can
+			 * skip adding it.
+			 */
+			if (includeLine != NULL && (includeLine == currentHbaContents ||
+										includeLine[-1] == '\n'))
+			{
+				log_debug("Line already exists in %s, skipping %s",
+						  hbaFilePath, hbaLineBuffer->data);
+
+				continue;
```

Comment:
On this `continue` and the one below we leak `hbaLineReplicationBuffer` and `hbaLineDatabaseBuffer`. I think it would be easiest and least error-prone to future changes to define the variables outside of the loop and free them at the start before allocating.

Reply:
Fixed, thanks.